File: 16_June_23/1569.java
// Intuition
// The problem asks for the number of different ways to reorder the given array nums such that the constructed Binary Search Tree (BST) is identical to the original BST formed from nums. We can solve this problem recursively by considering the root element of the BST and splitting the remaining elements into two groups: the left subtree elements (less than the root) and the right subtree elements (greater than the root). The total number of ways to reorder nums can be calculated by multiplying the number of ways to reorder the left subtree, the number of ways to reorder the right subtree, and the number of possible combinations of left and right subtrees.

// Approach
// Create a helper function countWays that takes a list of integers (nums) as input and returns the number of ways to reorder nums to form the same BST.
// In the countWays function:
// If the size of nums is less than or equal to 2, return 1 since there is only one way to reorder 0 or 1 element.
// Create two empty lists: left and right.
// Set the first element of nums as the root value.
// Iterate through the remaining elements of nums (starting from the second element):
// If the current element is less than the root value, add it to the left list.
// Otherwise, add it to the right list.
// Calculate the number of ways to reorder the left list (recursively) and store it in leftCount.
// Calculate the number of ways to reorder the right list (recursively) and store it in rightCount.
// Calculate the number of possible combinations of left and right subtrees using the comb function.
// Return the product of comb, leftCount, and rightCount.
// Create a numOfWays function that converts the input array into a list and calls the countWays function. Subtract 1 from the result and return it as the final output.
// Test the program with provided test cases or additional test cases.
// Complexity
// Time complexity:
// The time complexity of the solution is O(n^2), where n is the size of the nums array. This is because for each recursive call, we split the list into left and right subtrees, and in the worst case, the number of elements in the subtrees can be O(n). The comb function also has a time complexity of O(n^2) because it calculates combinations using dynamic programming.

// Space complexity:
// The space complexity is O(n) because we use additional space to store the left and right subtrees. The depth of the recursion can be at most n, so the space required for recursion is also O(n).

class Solution {
    private static final int MOD = 1000000007;
    public int numOfWays(int[] nums) {
        List<Integer> list = new ArrayList<>();
        for (int num : nums) {
            list.add(num);
        }
        return countWays(list) - 1;
    }
    private int countWays(List<Integer> nums) {
        if (nums.size() <= 2) {
            return 1;
        }
        
        List<Integer> left = new ArrayList<>();
        List<Integer> right = new ArrayList<>();
        int root = nums.get(0);
        
        for (int i = 1; i < nums.size(); i++) {
            if (nums.get(i) < root) {
                left.add(nums.get(i));
            } else {
                right.add(nums.get(i));
            }
        }
        
        long leftCount = countWays(left);
        long rightCount = countWays(right);
        
        return (int) ((comb(nums.size() - 1, left.size()) % MOD) * (leftCount % MOD) % MOD * (rightCount % MOD) % MOD);
    }
    
    private long comb(int n, int k) {
        long[][] dp = new long[n + 1][k + 1];
        for (int i = 0; i <= n; i++) {
            dp[i][0] = 1;
            for (int j = 1; j <= Math.min(i, k); j++) {
                dp[i][j] = (dp[i - 1][j - 1] + dp[i - 1][j]) % MOD;
            }
        }
        return dp[n][k];
    }
}

File: June_2023/16_June_23/1569.py
# Intuition
# Here are my initial thoughts on how to solve this problem:

# The base case of the recursive function f is when the length of the input list nums is less than or equal to 2, in which case there is only one way to arrange the numbers. This is returned as 1.

# The recursive case involves splitting the list nums into two sublists: left and right. The left sublist contains numbers smaller than the first element of nums, while the right sublist contains numbers greater than the first element.

# The number of ways to arrange the numbers in nums is calculated as the product of three factors:

# The number of ways to choose the positions for the right sublist among the total number of positions (i.e., combinations of the lengths of left and right).

# The number of ways to arrange the numbers in the left sublist.

# The number of ways to arrange the numbers in the right sublist.

# The recursive calls to f are made on the left and right sublists to compute the number of ways to arrange their respective numbers.

# Finally, the result is obtained by subtracting 1 from the total number of ways and taking the modulus with (10**9 + 7).

# Approach
# The numOfWays function calculates the number of ways to arrange the given list of numbers by recursively calling the helper function f.

# The recursive function f performs the following steps:

# It checks for the base case: if the length of the input list nums is less than or equal to 2, indicating that there is only one way to arrange the numbers, it returns 1.

# For the recursive case, the function splits the input list nums into two sublists: left and right. The left sublist contains all the numbers smaller than the first element of nums, while the right sublist contains all the numbers greater than the first element.

# The number of ways to arrange the numbers in nums is calculated as follows:

# Determine the number of positions available for the right sublist by combining the lengths of the left and right sublists using the comb function from the math module.

# Multiply the above value by the number of ways to arrange the numbers in the left sublist (by making a recursive call to f).

# Multiply the result by the number of ways to arrange the numbers in the right sublist (also by making a recursive call to f).

# The final result is obtained by subtracting 1 from the total number of ways calculated and taking the modulus with (10**9 + 7).

# To improve the code:
# Import the comb function from the math module.
# Use proper formatting and indentation to enhance code readability.
# Testing the code with different inputs and comparing the results against expected outcomes would be crucial to ensure the correctness of the implementation.

# Complexity
# Time complexity:O(2n)O(2^n)O(2 
# n
#  )
# Space complexity:O(n)O(n)O(n)

from math import comb
from typing import List

class Solution:
    def numOfWays(self, nums: List[int]) -> int:
        def f(nums):
            if len(nums) <= 2:
                return 1
            left = [v for v in nums if v < nums[0]]
            right = [v for v in nums if v > nums[0]]
            return comb(len(left) + len(right), len(right)) * f(left) * f(right)
        
        return (f(nums) - 1) % (10**9 + 7)